controllers/metal3.io/bmceventsubscription_controller.go

```diff
@@ -0,0 +1,278 @@
+			// garbage collected. For additional cleanup logic use
+			// finalizers.  Return and don't requeue
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, errors.Wrap(err, "could not load host data")
+	}
+
+	// Add a finalizer to newly created objects.
+	if subscription.DeletionTimestamp.IsZero() && !subscriptionHasFinalizer(subscription) {
+		reqLogger.Info(
+			"adding finalizer",
+			"existingFinalizers", subscription.Finalizers,
+			"newValue", metal3v1alpha1.BMCEventSubscriptionFinalizer,
+		)
+		subscription.Finalizers = append(subscription.Finalizers,
+			metal3v1alpha1.BMCEventSubscriptionFinalizer)
+		err := r.Update(ctx, subscription)
+		if err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to add finalizer")
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// Retrieve the BMC details from the host spec and validate host
+	// BMC details and build the credentials for talking to the
+	// management controller.
+	var bmcCreds *bmc.Credentials
+	var bmcCredsSecret *corev1.Secret
+
+	bmcCreds, bmcCredsSecret, err = r.buildAndValidateBMCCredentials(request, host)
+	if err != nil || bmcCreds == nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to get bmc secret")
+	}
+
+	initialState := host.Status.Provisioning.State
+	info := &reconcileInfo{
+		log:            reqLogger.WithValues("provisioningState", initialState),
+		host:           host,
+		request:        request,
+		bmcCredsSecret: bmcCredsSecret,
+	}
+
+	prov, err := r.ProvisionerFactory.NewProvisioner(provisioner.BuildHostData(*host, *bmcCreds), info.publishEvent)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to create provisioner")
+	}
+
+	ready, err := prov.IsReady()
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to check services availability")
+	}
+	if !ready {
+		reqLogger.Info("provisioner is not ready", "RequeueAfter:", provisionerNotReadyRetryDelay)
+		return ctrl.Result{Requeue: true, RequeueAfter: provisionerNotReadyRetryDelay}, nil
+	}
+
+	httpHeaders, err := r.getHTTPHeaders(*subscription)
+
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to get http headers")
+	}
+
+	if subscription.DeletionTimestamp.IsZero() {
+		// Not being deleted
+		if subscription.Status.SubscriptionID == "" {
+			if _, err := prov.AddBMCEventSubscriptionForNode(subscription, httpHeaders); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to add a subscription")
+			}
+
+			// Update the subscription with the new ID
+			err := r.Status().Update(ctx, subscription)
+			if err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to update subscription status")
+			}
+
+		} else {
+			// Update an existing subscription by recreating it in the BMC
+
+			originalSubscription := subscription.DeepCopy()
+
+			if _, err := prov.AddBMCEventSubscriptionForNode(subscription, httpHeaders); err != nil {
```

Comment:
Q: Are all the update eligible to destroy/recreate the subscription?

Reply:
Maybe @iurygregory knows?

Reply:
@andfasano  We can only create/delete/list a subscription, update is not supported (only in very newer redfish versions that most of the HW vendors are not using in their BMCs)

Reply:
I made them immutable. It simplifies the business logic in the reconciler while the webhook prevents updates.